Upgrade documentation to use Sphinx 2

Caster opened the pull request:
This pull request updates the documentation to work with Sphinx 2. It reorganizes the documentation a bit, moving images to one directory for example. It also gets rid of most of the custom Python scripts. Two remain, in the `_extensions`, and are called from Sphinx' `conf.py`. These scripts are responsible for generating pages from the help of tools, and generating pages from Doxygen output. Since there are no built-in capabilities in Sphinx for these tasks, it seems acceptable to leave these scripts.

Instead of CMake passing the mCRL2 version as an argument to `generate.py`, now there is some code in `conf.py` that will execute `MCRL2Version.CMake` in script mode to determine the version of mCRL2. Sphinx hence is still aware of the current mCRL2 version.

Caster commented:
I chose not to squash my commits to maintain the history and maybe make this easier to review, but it would probably be cleaner if you squash when merging. Of course, I leave that up to you.

tneele commented on tools/experimental/CMakeLists.txt:
```diff
@@ -1,5 +1,6 @@
 set(MCRL2_TOOLS
   besconvert  
+  complps2pbes
```
This is probably not a good idea. There's a reason this experimental tool is not included in the targets. Does it even compile?

Caster replied:
Okay, this is related to me not having the documentation for pbesabstract. I compiled mCRL2 locally and it did not compile the complps2pbes and pbesabstract tools. Because the documentation build process scans the `src/tools/experimental` directory to determine which tools to generate help pages for, it does not have help for these tools.

My solution for fixing the missing complps2pbes man page was to add it here. Is there a better way to compile these two tools? Some CMake flag I have overlooked?

tneele replied:
I think it would be best to imitate the behaviour of the old scripts: scan the `tools/{experimental,release}` directories and try to run each tool to obtain a man page. If a tool cannot be run and no RST file exists for it, no tool page is generated. Is this solution feasible?

Caster replied:
Yes, that's the current approach. I will revert this specific change.

Caster replied:
I do wonder though - how does the website have tool documentation for these tools currently? They must be compiled somewhere. Or is that cached?

Caster replied:
Reverted this change in 7fe4bdca279cf79af15851b93ec9fd982855a5ea.

tneele replied:
> How does the website have tool documentation for these tools currently?
I don't see the pages you are referring to. On https://mcrl2.org/web/user_manual/tools.html, there is no link for complps2pbes and the page for pbesabstract does not contain a man page.

Caster commented on doc/sphinx/Makefile:
```diff
@@ -0,0 +1,22 @@
+# Minimal makefile for Sphinx documentation
```
What I currently did is integrate this Makefile with the CMake build system by calling it from CMake. This file was generated automatically by Sphinx (`sphinx-quickstart` to be precise). If I understand you correctly, you suggest moving the Makefile into the CMakeLists.txt, right?

tneele replied:
Yes, that's right. At the moment, GNU Make is not required to run a build. One is free to use one of the other build systems, such as Ninja or nmake. Keeping the whole build system in CMake helps platform independence. This Makefile is not very complicated, so it should be doable to port the code to CMakeLists.txt.

Caster replied:
Done in da8b2cbf8b4de9067033e4cfa4fbfa2db64ea241.

Caster commented on .gitignore:
```diff
@@ -1,5 +1,8 @@
 build/SourceVersion
 
+# Generated by our build system
+doc/sphinx/html/mathjax/
```
I can in fact drop this line, because we no longer generate this directory. I may have tested this incorrectly before, but I thought the docs were always build in the source directory. That definitely is the case now (`_build` directory in `doc/sphinx`). What would be the preferred location for outputting the build?

tneele replied:
Yes. I typically create a clean directory (outside of the source tree), and then run something like
```
cmake ../src
make
```
Up to now, this would not create or change any files in the source directory. Is that also the case after your changes?

Caster replied:
Sphinx now always builds in the `_build` directory, which is in the sources. Furthermore, one of the custom extensions may add include directives to the tool pages if they are not there yet.

I will look into having Sphinx build in another directory, relative to the root CMake call.

mlaveaux replied:
You can use ${CMAKE_CURRENT_BINARY_DIR} to get a path to the out-of-source build directory relative to the CMakeLists.txt file (in Thomas' case ../src/doc/). Then the sphinx build directory would be ${CMAKE_CURRENT_BINARY_DIR}/_build.

Caster replied:
That's exactly what I did now, @mlaveaux. Except I renamed the directory to `sphinx` instead of `_build`, that makes it easier to identify what it is (and CMake already created a `doc` directory). Just have to correct the custom Python code, which still writes to `_build`.

Caster replied:
Fixed in b03067e778612b01daefc9c944ec7c993a78fb95 and 9ebcdcca3652c1b88707fe38f944a619be81e5e3.

Caster commented on build/cmake/MCRL2Version.cmake:
```diff
@@ -91,8 +91,13 @@ else()
 
-if("${CMAKE_CFG_INTDIR}" STREQUAL ".")
-  add_definitions(-DMCRL2_BUILD_TYPE=${CMAKE_BUILD_TYPE})
+if(CMAKE_SCRIPT_MODE_FILE AND CMAKE_SCRIPT_MODE_FILE STREQUAL CMAKE_PARENT_LIST_FILE)
```
Fixed in f1886b837602e4ad86e9fd8e7d0db6044c5f5707.

Caster commented:
Thank you for the review! I'll be working on resolving comments in the next hour or so.

> I was not able to place a comment for the tool docs of pbesabstract: did you forget to include the man page there?

The man page includes are placed there automatically by the custom extensions. It did it in a temporary location before, but that can no longer be done because we run from inside the Sphinx build process. So now I modified the extension to check whether the include is already present, and if not, it adds it. The include for pbesabstract is not placed because no help is generated for that tool. I'll investigate why, I just saw on the current page that there should be a help.

Caster commented on doc/sphinx/user_manual/tools/experimental/complps2pbes.rst:
```diff
@@ -0,0 +1,8 @@
+complps2pbes
+============
+
+.. include:: man/complps2pbes.txt
```
As I explained in a comment on the main thread, this is actually generated automatically by the custom extension. It just does it in the source now, which is tracked by Git. I felt that this is an okay way to do it, but I'm open to changing it.

tneele replied:
I see now. As also noted in another comment, I think it's nicest to leave the source tree untouched when building documentation.

Caster replied:
The only way I see to do that is to copy the entire source tree, make modifications to some of the files (man pages) and finally run Sphinx on that copy. That seems overkill, but it is definitely possible. Would that be preferable?

tneele replied:
Maybe a good intermediate solution is to only copy the rst files related to tool files to the build directory?

Caster replied:
Copying only part of the files is not workable, because Sphinx scans an entire directory (and only one) for source files. I have fixed this by copying the entire source tree (see 65f3910356974449f363c4f3f063955a7829a6c4 and following commits).

Caster commented on doc/sphinx/conf.py:
```diff
@@ -0,0 +1,128 @@
+version = f'{release}.{version}'
+
+# update copyright from current version
+copyright = f'2011-{version[0:4]}, Technische Universiteit Eindhoven'
```
True - it's easy to change this to use the current year. I copied this from [MCRL2Version.cmake](https://github.com/mCRL2org/mCRL2/blob/master/build/cmake/MCRL2Version.cmake#L28) - which I now realize I could also obtain from that file via the script mode. What would be preferable?

tneele replied:
Using the current year is the preferable solution, in my opinion.

Caster replied:
Done in 49b86c5a1337bbbe5d74526a737324cf80cad997.

Caster commented on doc/sphinx/README.md:
```diff
@@ -0,0 +1,44 @@
+equivalent path for your Python version), line 4261, which contains
+`assert len(withDecl) <= 1`: comment that out. It will error out unnecessarily.
+This is a [known issue](https://github.com/sphinx-doc/sphinx/issues/5496) in
+Sphinx.
```
The "proper workaround" would be to fork Sphinx and fix it there. The proper solution would be to open an issue, but then we probably have to make an MWE. I'll investigate.

Caster replied:
I avoided the issue by changing the reST output from Doxygen: 1183782147a558813b1f4e55eeb4b76b46d67412. What do you think?

tneele replied:
Looks great! :grin:

Caster replied:
Great! I've updated the README in b28fcba7c07cd6aaebaf916a61a6abb496f163fc.

Valo13 commented on doc/sphinx/user_manual/publications.rst:
```diff
@@ -86,7 +86,7 @@ Applications
 ---------------
 
-.. [GM14duplicate] J.F. Groote, M.R. Mousavi. `Modeling and analysis of communicating systems <https://mitpress.mit.edu/books/modeling-and-analysis-communicating-systems>`__. The MIT press. 2014.
+.. [GM14b] J.F. Groote, M.R. Mousavi. `Modeling and analysis of communicating systems <https://mitpress.mit.edu/books/modeling-and-analysis-communicating-systems>`__. The MIT press. 2014.
```
[GM14a] and [GM14b] are exactly the same book, wouldn't it be better to use [GM14] for both mentions? This also applies to `doc/sphinx/user_manual/tools/release/lps2lts.rst`

Caster replied:
Fixed in 7f98c77802bf44159255d772870a9d9aafcf7c03.

Valo13 commented on doc/sphinx/developer_manual/libraries/utilities/utilities_library.rst:
```diff
@@ -11,7 +11,7 @@ Structure
 The header files of the utilities library are roughly organised as depicted below.
 
-.. figure:: img/layout.png
+.. figure:: ../../../_static/utilities/layout.png
```
Wouldn't it be better to store this layout.png in the same folder or a subfolder (as it was before) to keep filepaths short and keep things together that belong together? If there is a considerate amount op reuse of such files then I understand why you would want these files in a separate place. This comment is applicable to more files.

mlaveaux replied:
You might be able to solve this using /_static/utilities/layout.png, where the first / indicates that the path is relative to the conf.py according to the answer given here: https://stackoverflow.com/questions/8475865/in-python-sphinx-how-to-link-to-a-file-in-static-directory

Caster replied:
The main reason for moving the images was to have a separation of images and text, much like ideally we separate source and build files. I shortened the paths as suggested by @mlaveaux in 71dd4fee3ab612ef93171b0f338e683acf853464 and hope that resolves this comment.

tneele commented:
Yes, I agree. It's also not completely clear what the license of `FileInformation.cmake` means for us. I think we are still safe, since a copy is not included in the binary releases.

Caster commented:
I have in bfc1eb111cea2a4de9f44598a6e6feee0b630bbf reverted the changes from 70668744cffa03457f5cdefe36b9fe6e173ba5ce.

Caster commented on doc/sphinx/developer_manual/libraries/.gitignore:
```diff
@@ -0,0 +1,37 @@
+!trace/trace_library.rst
+
+utilities/
+!utilities/utilities_library.rst
```
Fixed by 65f3910356974449f363c4f3f063955a7829a6c4 and following commits.

Caster commented:
With these last commits, I believe I have addressed all of the issues raised by @tneele, @Valo13 and @mlaveaux. Thank you again for your reviews and constructive feedback! Let me know if there is anything else I can do or help with.

Caster commented:
I have just rebase my branch onto master to ease merging.

mlaveaux commented:
All our comments seem to be addressed nicely and this project has been completed successfully. We have decided to upgrade our python dependency to version 3 for the whole configuration such that remains convenient to set up. After we have ensured that everything still works correctly we will integrate this pull request shortly after.
